R/buttons.R

```diff
@@ -0,0 +1,196 @@
+#'   in ready state. See [fontawesome::fa_i()].
+#' @param busy_label The label of the button while it is busy.
+#' @param busy_icon The icon to display while the button is busy. By default, a
+#'   spinning "chasing arrows" icon is used.
+#' @param type One of the Bootstrap theme colors (`"primary"`, `"default"`,
+#'   `"secondary"`, `"success"`, `"danger"`, `"warning"`, `"info"`, `"light"`,
+#'   `"dark"`), or `NULL` to leave off the Bootstrap-specific button CSS classes
+#'   altogether.
+#' @param ... Named arguments become attributes to include on the `<button>`
+#'   element.
+#' @param auto_reset If `TRUE` (the default), automatically put the button back
+#'   in "ready" state after its click is handled by the server.
+#'
+#' @section Server value:
+#' An integer of class `"shinyActionButtonValue"`. This class differs from
+#' ordinary integers in that a value of 0 is considered "falsy".
+#' This implies two things:
+#'   * Event handlers (e.g., [shiny::observeEvent()], [shiny::eventReactive()]) won't execute on initial load.
+#'   * Input validation (e.g., [shiny::req()], [shiny::need()]) will fail on initial load.
+#'
+#' @examples
+#' if (interactive()) {
+#'   library(shiny)
+#'   library(bslib)
+#'
+#'   ui <- page_sidebar(
+#'     sidebar = sidebar(open = "always",
+#'       input_task_button("resample", "Resample"),
+#'     ),
+#'     verbatimTextOutput("summary")
+#'   )
+#'
+#'   server <- function(input, output, session) {
+#'     sample <- eventReactive(input$resample, ignoreNULL=FALSE, {
+#'       Sys.sleep(2)  # Make this artificially slow
+#'       rnorm(100)
+#'     })
+#'
+#'     output$summary <- renderPrint({
+#'       summary(sample())
+#'     })
+#'   }
+#'
+#'   shinyApp(ui, server)
+#' }
+#'
+#' @export
+input_task_button <- function(id, label, icon = NULL,
+  busy_label = "Processing...", busy_icon = icon_spinner(), ...,
+  type = "primary", auto_reset = TRUE) {
+
+  tags$button(
+    id = id,
+    class = if (!is.null(type)) paste0("btn btn-", type),
+    class = "bslib-task-button",
+    type = "button",
+    "data-auto-reset" = if (isTRUE(auto_reset)) NA else NULL,
+
+    component_dependencies(),
+
+    htmltools::tag("bslib-switch-inline",
+      list(
+        case = "ready",
+        span(slot = "ready",
+          icon, label
+        ),
+        span(slot = "busy",
+          busy_icon, busy_label
+        )
+      )
+    ),
+
+    ...
+  )
+}
+
+#' @rdname input_task_button
+#' @export
+icon_spinner <- function() {
+  fontawesome::fa_i("refresh", class = "fa-spin", "aria-hidden" = "true")
+}
+
+#' @param state If `"busy"`, put the button into busy/disabled state. If
+#'   `"ready"`, put the button into ready/enabled state.
+#' @param session The `session` object; using the default is recommended.
+#' @rdname input_task_button
+#' @export
+update_task_button <- function(id, state = c(NULL, "ready", "busy"), session = get_current_session()) {
```

Comment:
I included a test for `missing(state)` below and no-op in that case. That said, I felt weird writing `c(NULL, "ready", "busy")`. I don't want to default to `"ready"`, if the button is in the busy state and (in the future) you used this function to change the label or something, it would be weird to change the state to `ready` when it wasn't asked for.

The other possibility I considered is `state = NULL`, which will (hopefully) clearly signal that it no-ops by default, but then you have to look at the docs to see that `"ready"` and `"busy"` are the two options.

I could also do `c("", "ready", "busy")` but that feels weird too--I don't think we have `""` as an `updateXXX` no-op anywhere else, it's always `NULL`.

Reply:
I agree with you about the weirdness you've identified. Of all the options, I like this one the best:

```suggestion
update_task_button <- function(id, state, ..., session = get_current_session()) {
```

1. `state` as a required argument because you _need_ to have a state or it doesn't make sense to call `update_task_button()`.
2. Add `...` and move `session` behind them so we leave room in our function signature for future expansion, knowing that we may add other properties in the future.

Reply:
After discussion: `state = NULL`, add `...` to come before `state`.